Based on the OpenBSD ports patch:
https://github.com/openbsd/ports/tree/c1521c4eb969ddf29306117ea125802a2644e734/www/h2o/patches

Make neverbleed work without the (deprecated, removed in OpenSSL 4.0)
ENGINE API. When OPENSSL_NO_ENGINE is defined, store RSA_METHOD /
EC_KEY_METHOD in neverbleed_t and attach them with RSA_set_method() /
EC_KEY_set_method() instead of registering an ENGINE. The QAT offload
support is disabled in that case, as it requires the ENGINE API.

Rebased onto the current H2O master (2026-06, neverbleed a191b3c).

--- a/deps/neverbleed/neverbleed.c
+++ b/deps/neverbleed/neverbleed.c
@@ -80,7 +80,8 @@
 #endif
 
 #ifdef __linux
-#if OPENSSL_VERSION_NUMBER >= 0x1010000fL && !defined(LIBRESSL_VERSION_NUMBER) && !defined(OPENSSL_IS_BORINGSSL)
+#if OPENSSL_VERSION_NUMBER >= 0x1010000fL && !defined(LIBRESSL_VERSION_NUMBER) && !defined(OPENSSL_IS_BORINGSSL) &&             \
+    !defined(OPENSSL_NO_ENGINE)
 #define USE_OFFLOAD 1
 #endif
 #if defined(OPENSSL_IS_BORINGSSL) && defined(NEVERBLEED_BORINGSSL_USE_QAT)
@@ -915,7 +916,15 @@
     exdata->nb = nb;
     exdata->key_index = key_index;
 
+#ifdef OPENSSL_NO_ENGINE
+    if ((rsa = RSA_new()) == NULL) {
+        fprintf(stderr, "no memory\n");
+        abort();
+    }
+    RSA_set_method(rsa, nb->rsa_method);
+#else
     rsa = RSA_new_method(nb->engine);
+#endif
     RSA_set_ex_data(rsa, get_rsa_exdata_idx(), exdata);
     if (BN_hex2bn(&e, ebuf) == 0) {
         fprintf(stderr, "failed to parse e:%s\n", ebuf);
@@ -1063,7 +1072,15 @@
     exdata->nb = nb;
     exdata->key_index = key_index;
 
+#ifdef OPENSSL_NO_ENGINE
+    if ((ec_key = EC_KEY_new()) == NULL) {
+        fprintf(stderr, "no memory\n");
+        abort();
+    }
+    EC_KEY_set_method(ec_key, nb->ecdsa_method);
+#else
     ec_key = EC_KEY_new_method(nb->engine);
+#endif
     EC_KEY_set_ex_data(ec_key, get_ecdsa_exdata_idx(), exdata);
 
     ec_group = EC_GROUP_new_by_curve_name(curve_name);
@@ -2261,6 +2278,14 @@
         EC_KEY_METHOD_set_sign(ecdsa_method, ecdsa_sign_proxy, NULL, NULL);
 #endif
 
+#ifdef OPENSSL_NO_ENGINE
+        nb->rsa_method = rsa_method;
+#ifdef NEVERBLEED_ECDSA
+        nb->ecdsa_method = ecdsa_method;
+#else
+        nb->ecdsa_method = NULL;
+#endif
+#else
         if ((nb->engine = ENGINE_new()) == NULL || !ENGINE_set_id(nb->engine, "neverbleed") ||
             !ENGINE_set_name(nb->engine, "privilege separation software engine") || !ENGINE_set_RSA(nb->engine, rsa_method)
 #ifdef NEVERBLEED_ECDSA
@@ -2271,6 +2296,7 @@
             goto Fail;
         }
         ENGINE_add(nb->engine);
+#endif
     }
 #endif
 
@@ -2290,10 +2316,12 @@
     }
     if (listen_fd != -1)
         close(listen_fd);
+#ifndef OPENSSL_NO_ENGINE
     if (nb->engine != NULL) {
         ENGINE_free(nb->engine);
         nb->engine = NULL;
     }
+#endif
     return -1;
 }
 
--- a/deps/neverbleed/neverbleed.h
+++ b/deps/neverbleed/neverbleed.h
@@ -24,7 +24,10 @@
 
 #include <pthread.h>
 #include <sys/un.h>
+#include <openssl/ssl.h>
+#ifndef OPENSSL_NO_ENGINE
 #include <openssl/engine.h>
+#endif
 
 #ifdef __FreeBSD__
 #include <pthread_np.h>
@@ -46,7 +49,12 @@
 #define NEVERBLEED_AUTH_TOKEN_SIZE 32
 
 typedef struct st_neverbleed_t {
+#ifdef OPENSSL_NO_ENGINE
+    RSA_METHOD *rsa_method;
+    void *ecdsa_method;
+#else
     ENGINE *engine;
+#endif
     pid_t daemon_pid;
     struct sockaddr_un sun_;
     pthread_key_t thread_key;
